lib/ain-evm/src/filters.rs

```diff
@@ -0,0 +1,206 @@
+                topics,
+                from_block,
+                to_block,
+                last_block_height: current_block_height,
+            }),
+        );
+
+        return *filter_id;
+    }
+
+    pub fn create_block_filter(&self) -> usize {
+        let mut filters = self.filters.write().unwrap();
+        let mut filter_id = self.filter_id.write().unwrap();
+        *filter_id += 1;
+
+        let filter = Filter::NewBlock(BlockFilter {
+            id: *filter_id,
+            block_hashes: vec![],
+        });
+
+        filters.insert(*filter_id, filter);
+
+        return *filter_id;
+    }
+
+    pub fn create_transaction_filter(&self) -> usize {
+        let mut filters = self.filters.write().unwrap();
+        let mut filter_id = self.filter_id.write().unwrap();
+        *filter_id += 1;
+
+        let filter = Filter::NewPendingTransactions(PendingTransactionFilter {
+            id: *filter_id,
+            transaction_hashes: vec![],
+        });
+
+        filters.insert(*filter_id, filter);
+
+        return *filter_id;
+    }
+
+    pub fn get_filter(&self, filter_id: usize) -> Result<Filter, &str> {
+        let filters = self.filters.read().unwrap();
+
+        match filters.get(&filter_id) {
+            Some(filter) => Ok(filter.clone()),
+            None => Err("Unable to find filter"),
+        }
+    }
+
+    pub fn update_last_block(&self, filter_id: usize, block_height: U256) {
+        let mut filters = self.filters.write().unwrap();
+
+        let filter = filters.get_mut(&filter_id).unwrap();
+
+        match filter {
+            Filter::Logs(f) => f.last_block_height = block_height,
+            _ => {}
+        }
+    }
+
+    pub fn add_block_to_filters(&self, block_hash: H256) {
+        let mut filters = self.filters.write().unwrap();
+        debug!("Adding {:#x} to filters", block_hash);
+
+        for item in filters.iter_mut() {
+            let filter = item.1;
+            match filter {
+                Filter::NewBlock(filter) => {
+                    debug!("Added block hash to {:#?}", filter.clone());
+                    filter.block_hashes.push(block_hash);
+                }
+                _ => {}
+            }
+        }
+    }
+
+    pub fn add_tx_to_filters(&self, tx_hash: H256) {
+        let mut filters = self.filters.write().unwrap();
+        debug!("Adding {:#x} to filters", tx_hash);
+
+        for item in filters.iter_mut() {
+            let filter = item.1;
+            match filter {
+                Filter::NewPendingTransactions(filter) => {
+                    debug!("Added tx hash to {:#?}", filter.clone());
+                    filter.transaction_hashes.push(tx_hash);
+                }
+                _ => {}
+            }
+        }
+    }
+
+    pub fn get_entries_from_filter(&self, filter_id: usize) -> Result<Vec<H256>, &str> {
+        let mut filters = self.filters.write().unwrap();
+
+        let filter = filters.get_mut(&filter_id).unwrap();
```

Comment:
```suggestion
        let Some(filter) = filters.get_mut(&filter_id) else {
            return Err("Filter not found");
        };
```

Reply:
These should be safe to use since we get the filter to check if it exists at the beginning of every RPC call. I have added in the changes in case we use these methods elsewhere.